parser/json/parser_benchmark_test.go

```diff
@@ -0,0 +1,326 @@
+			}
+		}
+	})
+}
+
+func benchmarkFastJSONParseGet(b *testing.B, s string) {
+	b.ReportAllocs()
+	b.SetBytes(int64(len(s)))
+	b.RunParallel(func(pb *testing.PB) {
+		p := NewParser(NewOptions())
+		var n int
+		for pb.Next() {
+			v, err := p.Parse(s)
+			if err != nil {
+				panic(fmt.Errorf("unexpected error: %s", err))
+			}
+			nv, found := v.Get("sid")
+			if found && nv.Type() == value.NumberType {
+				n += int(nv.MustNumber())
+			}
+			sv, found := v.Get("uuid")
+			if found && sv.Type() == value.StringType {
+				n += len(sv.MustString())
+			}
+			p, _ := v.Get("person")
+			if p != nil {
+				n++
+			}
+			c, _ := v.Get("company")
+			if c != nil {
+				n++
+			}
+			u, _ := v.Get("users")
+			if u != nil {
+				n++
+			}
+			av, found := v.Get("features")
+			if found && av.Type() == value.ArrayType {
+				n += av.MustArray().Len()
+			}
+			av, found = v.Get("topicSubTopics")
+			if found && av.Type() == value.ArrayType {
+				n += av.MustArray().Len()
+			}
+			o, _ := v.Get("search_metadata")
+			if o != nil {
+				n++
+			}
+		}
+	})
+}
+
+func benchmarkStdJSONParseMap(b *testing.B, s string) {
+	b.ReportAllocs()
+	b.SetBytes(int64(len(s)))
+	bb := s2b(s)
+	b.RunParallel(func(pb *testing.PB) {
+		var m map[string]interface{}
+		for pb.Next() {
+			if err := json.Unmarshal(bb, &m); err != nil {
+				panic(fmt.Errorf("unexpected error: %s", err))
+			}
+		}
+	})
+}
+
+func benchmarkStdJSONParseStruct(b *testing.B, s string) {
+	b.ReportAllocs()
+	b.SetBytes(int64(len(s)))
+	bb := s2b(s)
+	b.RunParallel(func(pb *testing.PB) {
+		var m struct {
+			Sid            int
+			UUID           string
+			Person         map[string]interface{}
+			Company        map[string]interface{}
+			Users          []interface{}
+			Features       []map[string]interface{}
+			TopicSubTopics map[string]interface{}
+			SearchMetadata map[string]interface{}
+		}
+		for pb.Next() {
+			if err := json.Unmarshal(bb, &m); err != nil {
+				panic(fmt.Errorf("unexpected error: %s", err))
+			}
+		}
+	})
+}
+
+func benchmarkStdJSONParseEmptyStruct(b *testing.B, s string) {
+	b.ReportAllocs()
+	b.SetBytes(int64(len(s)))
+	bb := s2b(s)
+	b.RunParallel(func(pb *testing.PB) {
+		var m struct{}
+		for pb.Next() {
+			if err := json.Unmarshal(bb, &m); err != nil {
+				panic(fmt.Errorf("unexpected error: %s", err))
+			}
+		}
+	})
+}
+
+// NB(xichen): This should work in theory but have not been stress tested.
+// This is currently only called in tests.
+func s2b(s string) []byte {
+	if len(s) == 0 {
+		return nil
+	}
+
+	// NB(xichen): we need to declare a real byte slice so internally the compiler
+	// knows to use an unsafe.Pointer to keep track of the underlying memory so tha
+	// once the slice's array pointer is updated with the pointer to the string's
+	// underlying bytes, the compiler won't prematurely GC the memory when the string
+	// goes out of scope.
+	var b []byte
+	byteHeader := (*reflect.SliceHeader)(unsafe.Pointer(&b))
```

Comment:
So does casting to `reflect.TYPEHeader` give you access to the golang runtime's internal representation of a bye slice and a string?

Reply:
Yes.